incubator/hnc/pkg/validators/hncconfig.go

```diff
@@ -0,0 +1,182 @@
+// allowing it to be more easily unit tested (ie without constructing a full admission.Request).
+func (c *HNCConfig) handle(ctx context.Context, inst *api.HNCConfiguration) admission.Response {
+	if inst.GetName() != "config" {
+		return deny(metav1.StatusReasonInvalid, fmt.Sprintf("Wrong singleton name: %s; the name should be 'config'", inst.GetName()))
+	}
+
+	roleExist := false
+	roleBindingExist := false
+	ts := gvkSet{}
+	for _, t := range inst.Spec.Types {
+		if rp := c.isTypeConfigured(t, ts); !rp.Allowed {
+			return rp
+		}
+
+		if c.isRBAC(t, "Role") {
+			if rp := c.validateRBAC(t.Mode, "Role"); !rp.Allowed {
+				return rp
+			}
+			roleExist = true
+		} else if c.isRBAC(t, "RoleBinding") {
+			if rp := c.validateRBAC(t.Mode, "RoleBinding"); !rp.Allowed {
+				return rp
+			}
+			roleBindingExist = true
+		} else {
+			if rp := c.validateType(ctx, t, ts); !rp.Allowed {
+				return rp
+			}
+		}
+	}
+	if !roleExist {
+		return deny(metav1.StatusReasonInvalid, "Configuration for Role is missing")
+	}
+	if !roleBindingExist {
+		return deny(metav1.StatusReasonInvalid, "Configuration for RoleBinding is missing")
+	}
+	return allow("")
+}
+
+// Validate if the configuration of a type already exists. Each type should only have one configuration.
+func (c *HNCConfig) isTypeConfigured(t api.TypeSynchronizationSpec, ts gvkSet) admission.Response {
+	gvk := schema.FromAPIVersionAndKind(t.APIVersion, t.Kind)
+	if exists := ts[gvk]; exists {
+		return deny(metav1.StatusReasonInvalid, fmt.Sprintf("Duplicate configurations for %s", gvk))
+	}
+	ts[gvk] = true
+	return allow("")
+}
+
+func (c *HNCConfig) isRBAC(t api.TypeSynchronizationSpec, kind string) bool {
+	return t.APIVersion == "rbac.authorization.k8s.io/v1" && t.Kind == kind
+}
+
+// The mode of Role and RoleBinding should be either unset or set to the propagate mode.
+func (c *HNCConfig) validateRBAC(mode api.SynchronizationMode, kind string) admission.Response {
+	if mode == api.Propagate || mode == "" {
+		return allow("")
+	}
+	return deny(metav1.StatusReasonInvalid, fmt.Sprintf("Invalid mode of %s; current mode: %s; expected mode %s", kind, mode, api.Propagate))
+}
+
+// validateType validates a non-RBAC type.
+func (c *HNCConfig) validateType(ctx context.Context, t api.TypeSynchronizationSpec, ts gvkSet) admission.Response {
+	gvk := schema.FromAPIVersionAndKind(t.APIVersion, t.Kind)
+
+	// Validate if the GVK exists in the apiserver.
+	exists, err := c.validator.Exists(ctx, gvk)
+	if !exists {
+		return deny(metav1.StatusReasonInvalid,
+			fmt.Sprintf("Cannot find the %s in the apiserver with error: %s", gvk, err.Error()))
+	}
+
+	// The mode of a type should be either unset or set to one of the supported modes.
+	if t.Mode != api.Propagate && t.Mode != api.Remove && t.Mode != api.Ignore && t.Mode != "" {
+		return deny(metav1.StatusReasonInvalid, fmt.Sprintf("Unrecognized mode '%s' for %s", t.Mode, gvk))
+	}
```

Comment:
Nit: maybe better to write this as a switch statement?

Reply:
Thanks for the suggestion! Changing to the switch should be more readable. I am thinking about updating this in the same PR that changes `StatusReasonInvalid` (see my comment above), if we are going to change it. Otherwise, I can sent out a small PR to change this to the switch statement.

Reply:
Whichever you prefer is fine by me :)

Reply:
Thanks for the suggestion! I changed it to `switch` in #525